launch/launch.go

```diff
@@ -7,146 +7,114 @@ import (
+	return c
+}
+
+func (c RawCommand) MarshalTOML() ([]byte, error) {
+	if c.PlatformAPI == nil {
+		return nil, fmt.Errorf("missing PlatformAPI while encoding RawCommand")
 	}
 
-	// This is the same as launch.Process and exists to allow us to toml.Decode inside of UnmarshalTOML
-	type pProcess Process
+	if c.PlatformAPI.AtLeast("0.10") {
+		buffer := &strings.Builder{}
+		// turn array into toml array
+		buffer.WriteString("[")
+		for i, entry := range c.Entries {
+			if i != 0 {
+				buffer.WriteString(", ")
+			}
+			buffer.WriteString(fmt.Sprintf("%q", entry))
+		}
+		buffer.WriteString("]")
+		return []byte(buffer.String()), nil
+	}
 
-	// unmarshal the common bits
-	newProcess := pProcess{}
-	md, err := toml.Decode(tomlString, &newProcess)
-	if err != nil {
-		return err
+	return []byte(fmt.Sprintf("\"%s\"", c.Entries[0])), nil
+}
+
+func (c RawCommand) MarshalJSON() ([]byte, error) {
+	if c.PlatformAPI == nil {
+		return nil, fmt.Errorf("missing PlatformAPI while encoding RawCommand")
 	}
 
-	// handle the process.command, which will differ based on APIs
-	var commandWasString bool
-	var commandString string
-	if err := md.PrimitiveDecode(newProcess.RawCommandValue, &commandString); err == nil {
-		commandWasString = true
-		newProcess.Command = []string{commandString}
+	if c.PlatformAPI.AtLeast("0.10") {
+		return json.Marshal(c.Entries)
 	}
 
-	if !commandWasString {
-		var command []string
-		if err := md.PrimitiveDecode(newProcess.RawCommandValue, &command); err != nil {
-			return err
+	return []byte(fmt.Sprintf("\"%s\"", c.Entries[0])), nil
+}
+
+// UnmarshalTOML implements toml.Unmarshaler and is needed because we read metadata.toml
+// this method will attempt to parse the command in either string or array format
+func (c *RawCommand) UnmarshalTOML(data interface{}) error {
+	var values []string
+	// the raw value is either "the-command" or ["the-command", "arg1", "arg2"]
+	// the latter is exposed as []interface{} by toml library and needs conversion
+	switch v := data.(type) {
+	case string:
+		values = []string{v}
+	case []interface{}:
+		s := make([]string, len(v))
+		for i, v := range v {
+			s[i] = fmt.Sprint(v)
 		}
-		newProcess.Command = command
+		values = s
+	default:
+		return fmt.Errorf("unknown command toml type %T", data)
 	}
 
-	*p = Process(newProcess)
+	*c = NewRawCommand(values)
 	return nil
 }
 
+// Process represents a process to launch at runtime.
+type Process struct {
+	Type             string       `toml:"type" json:"type"`
+	Command          RawCommand   `toml:"command" json:"command"`
+	Args             []string     `toml:"args" json:"args"`
+	Direct           bool         `toml:"direct" json:"direct"`
+	Default          bool         `toml:"default,omitempty" json:"default,omitempty"`
+	BuildpackID      string       `toml:"buildpack-id" json:"buildpackID"`
+	WorkingDirectory string       `toml:"working-dir,omitempty" json:"working-dir,omitempty"`
+	PlatformAPI      *api.Version `toml:"-" json:"-"`
+}
+
```

Comment:
nit: can we move this to the top of the file?

Reply:
[d3cc6ba](https://github.com/buildpacks/lifecycle/pull/946/commits/d3cc6baa6df958e47430b2f59090fc661fff3f81)